server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -388,6 +399,37 @@ private void failMissingShards(final ClusterState state) {
         }
     }
 
+    /**
+     * Removes the {@link IndexService} of indices whose state has changed.
+     * Closing the index services here will force them to be recreated later along with their shards.
+     */
+    private void removeIndices(final ClusterChangedEvent event) {
```

Comment:
can we fold this into `updateIndices`? There we use the same iteration order and have the same lookup patterns, so I think we can save a lot on the boiler plate and can avoid iterating yet another time over all indices